File: unity/Assets/Scripts/Multiplayer/Rooms/CurrentRoomInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;
using static Michitai.Multiplayer.Time.Time;

namespace Michitai.Multiplayer.Rooms
{
    /// <summary>
    /// Detailed information about the current game room.
    /// Contains room configuration, player counts, status, and rules.
    /// Uses Unity's JsonUtility for deserialization of rules.
    /// </summary>
    /// <typeparam name="T">The type to deserialize room rules into.</typeparam>
    [System.Serializable]
    public class CurrentRoomInfo<T> where T : class, new()
    {
        [SerializeField]
        private string joined_at;
        [SerializeField]
        private string last_heartbeat;
        [SerializeField]
        private string room_created_at;
        [SerializeField]
        private string room_last_activity;
        [SerializeField]
        private string rules_json;          // Unity mode

        /// <summary>
        /// The unique ID of the room.
        /// </summary>
        public string room_id;

        /// <summary>
        /// The name of the room.
        /// </summary>
        public string room_name;

        /// <summary>
        /// Whether the current player is the host.
        /// </summary>
        public bool is_host;

        /// <summary>
        /// Whether the room is currently online/active.
        /// </summary>
        public bool is_online;

        /// <summary>
        /// Maximum number of players allowed.
        /// </summary>
        public int max_players;

        /// <summary>
        /// Current number of players in the room.
        /// </summary>
        public int current_players;

        /// <summary>
        /// Whether the room has a password set.
        /// </summary>
        public bool has_password;

        /// <summary>
        /// Whether host switching is allowed.
        /// </summary>
        public bool host_switch;

        /// <summary>
        /// Whether players can leave the room.
        /// </summary>
        public bool can_leave;

        /// <summary>
        /// Whether the room supports realtime communication.
        /// </summary>
        public bool realtime;

        /// <summary>
        /// Whether the room is currently active.
        /// </summary>
        public bool is_active;

        /// <summary>
        /// The current player's name.
        /// </summary>
        public string player_name;

        /// <summary>
        /// The room rules deserialized into the specified type.
        /// </summary>
        public T Rules
        {
            get
            {
                return JsonUtility.FromJson<T>(rules_json);
            }
        }

        /// <summary>
        /// Timestamp when the current player joined the room.
        /// </summary>
        public DateTimeOffset? JoinedAt
        {
            get
            {
                return ParseUtc(joined_at);
            }
        }

        /// <summary>
        /// Timestamp of the last heartbeat from the current player.
        /// </summary>
        public DateTimeOffset? LastHeartbeat
        {
            get
            {
                return ParseUtc(last_heartbeat);
            }
        }

        /// <summary>
        /// Timestamp when the room was created.
        /// </summary>
        public DateTimeOffset? RoomCreatedAt
        {
            get
            {
                return ParseUtc(room_created_at);
            }
        }

        /// <summary>
        /// Timestamp of the last activity in the room.
        /// </summary>
        public DateTimeOffset? RoomLastActivity
        {
            get
            {
                return ParseUtc(room_last_activity);
            }
        }
    }
}

File: java/src/main/java/com/michitai/multiplayer/rooms/CurrentRoomInfo.java
package com.michitai.multiplayer.rooms;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Comprehensive information about a game room.
 *
 * @param <T> The type to deserialize room rules into.
 */
public class CurrentRoomInfo<T> {
    @JsonProperty("room_id")
    private String roomId;

    @JsonProperty("room_name")
    private String roomName;

    @JsonProperty("is_host")
    private boolean isHost;

    @JsonProperty("is_online")
    private boolean isOnline;

    @JsonProperty("max_players")
    private int maxPlayers;

    @JsonProperty("current_players")
    private int currentPlayers;

    @JsonProperty("has_password")
    private boolean hasPassword;

    @JsonProperty("host_switch")
    private boolean hostSwitch;

    @JsonProperty("can_leave")
    private boolean canLeave;

    @JsonProperty("realtime")
    private boolean realtime;

    @JsonProperty("is_active")
    private boolean isActive;

    @JsonProperty("rules")
    private T rules;

    @JsonProperty("player_name")
    private String playerName;

    @JsonProperty("joined_at")
    private String joinedAt;

    @JsonProperty("last_heartbeat")
    private String lastHeartbeat;

    @JsonProperty("room_created_at")
    private String roomCreatedAt;

    @JsonProperty("room_last_activity")
    private String roomLastActivity;

    public String getRoomId() {
        return roomId;
    }

    public void setRoomId(String roomId) {
        this.roomId = roomId;
    }

    public String getRoomName() {
        return roomName;
    }

    public void setRoomName(String roomName) {
        this.roomName = roomName;
    }

    public boolean isHost() {
        return isHost;
    }

    public void setHost(boolean host) {
        isHost = host;
    }

    public boolean isOnline() {
        return isOnline;
    }

    public void setOnline(boolean online) {
        isOnline = online;
    }

    public int getMaxPlayers() {
        return maxPlayers;
    }

    public void setMaxPlayers(int maxPlayers) {
        this.maxPlayers = maxPlayers;
    }

    public int getCurrentPlayers() {
        return currentPlayers;
    }

    public void setCurrentPlayers(int currentPlayers) {
        this.currentPlayers = currentPlayers;
    }

    public boolean isHasPassword() {
        return hasPassword;
    }

    public void setHasPassword(boolean hasPassword) {
        this.hasPassword = hasPassword;
    }

    public boolean isHostSwitch() {
        return hostSwitch;
    }

    public void setHostSwitch(boolean hostSwitch) {
        this.hostSwitch = hostSwitch;
    }

    public boolean isCanLeave() {
        return canLeave;
    }

    public void setCanLeave(boolean canLeave) {
        this.canLeave = canLeave;
    }

    public boolean isRealtime() {
        return realtime;
    }

    public void setRealtime(boolean realtime) {
        this.realtime = realtime;
    }

    public boolean isActive() {
        return isActive;
    }

    public void setActive(boolean active) {
        isActive = active;
    }

    public T getRules() {
        return rules;
    }

    public void setRules(T rules) {
        this.rules = rules;
    }

    public String getPlayerName() {
        return playerName;
    }

    public void setPlayerName(String playerName) {
        this.playerName = playerName;
    }

    public String getJoinedAt() {
        return joinedAt;
    }

    public void setJoinedAt(String joinedAt) {
        this.joinedAt = joinedAt;
    }

    public String getLastHeartbeat() {
        return lastHeartbeat;
    }

    public void setLastHeartbeat(String lastHeartbeat) {
        this.lastHeartbeat = lastHeartbeat;
    }

    public String getRoomCreatedAt() {
        return roomCreatedAt;
    }

    public void setRoomCreatedAt(String roomCreatedAt) {
        this.roomCreatedAt = roomCreatedAt;
    }

    public String getRoomLastActivity() {
        return roomLastActivity;
    }

    public void setRoomLastActivity(String roomLastActivity) {
        this.roomLastActivity = roomLastActivity;
    }
}
